### PR TITLE
Switch __count field in Enum to strum

### PR DESCRIPTION
**Pull request by AnhTuDo1998**
fix: Use strum crate for consistency

This PR is for problem #6520. It replaces ```__count``` field in ```ExtCosts``` & ```ActionCosts``` Enums with ```strum::EnumCount``` and change codes where count of fields in Enum is needed. As we used strum crates in a few place in our project, this change is meant for consistency.

Compilation and existing unit test or ```core/primitives-core``` passed

### CODE REVIEW

File: core/primitives-core/src/config.rs [hunk 1]
```diff
@@ -643,7 +638,7 @@ impl fmt::Display for ActionCosts {
 
 impl ActionCosts {
     pub const fn count() -> usize {
-        ActionCosts::__count as usize
+        ActionCosts::COUNT as usize
     }
```
Comment (matklad) [inline core/primitives-core/src/config.rs:642]: Let's remove this method altogether

File: core/primitives-core/src/config.rs [hunk 2]
```diff
@@ -738,13 +733,11 @@ impl ExtCosts {
             alt_bn128_g1_sum_base => config.alt_bn128_g1_sum_base,
             #[cfg(feature = "protocol_feature_alt_bn128")]
             alt_bn128_g1_sum_byte => config.alt_bn128_g1_sum_byte,
-
-            __count => unreachable!(),
         }
     }
 
     pub const fn count() -> usize {
```
Comment (matklad) [inline core/primitives-core/src/config.rs:739]: and this one

**Comment by AnhTuDo1998** [comment]
To test the changes in using ```strum::Display```, I have hardcoded locally a few ```println!()``` in unit test of profile.rs that previously use ```name_of()```. Here is a snippet of ```cargo test -- --nocapture```, let me know if this is alright

image

**Comment by matklad** [comment]
Yeah, this looks good!

**Review by matklad — APPROVED** [review]

File: core/primitives-core/src/config.rs [hunk 3]
```diff
@@ -611,13 +611,10 @@ pub enum ExtCosts {
...
     #[cfg(feature = "protocol_feature_alt_bn128")]
     alt_bn128_g1_sum_byte,
-
-    // NOTE: this should be the last element of the enum.
-    __count,
 }
 
 // Type of an action, used in fees logic.
-#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug, PartialOrd, Ord)]
+#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug, PartialOrd, Ord, EnumCount, Display)]
```
Comment (nagisa) [inline core/primitives-core/src/config.rs:617]: I'd argue that “Display” is way too generic of a name to use in `derive` without referring to it by a crate (i.e. `strum::Display`), but this is probably a discussion to be had outside of this PR.

Comment (matklad) [reply core/primitives-core/src/config.rs:617]: Uhu, same thoughts here

Comment (matklad) [reply core/primitives-core/src/config.rs:617]: I'd even say this is not specific to `Display`: `strum` is a minor crate, with a short name, so we should just `strum::EnumCount, strum::Display`, etc.

Comment (nagisa) [reply core/primitives-core/src/config.rs:617]: Yeah, I kinda prefer taking that approach to the extreme of “all external derives must be prefixed, and not imported", but either way this seems like a discussion best had over a changeset to the style guide.

**Pull request merged**